osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -35,147 +44,306 @@
+     *         DLMS device id
+     * @param key
+     *         key to store, unencrypted
+     * @param keyType
+     *         type of key
+     *
+     * @see #activateNewKey(String, SecurityKeyType)
+     */
+    public void storeNewKey(String deviceIdentification, SecurityKeyType keyType, byte[] key) {
+        Map<SecurityKeyType, byte[]> keysByType = new EnumMap<>(SecurityKeyType.class);
+        keysByType.put(keyType, key);
+        this.storeNewKeys(deviceIdentification, keysByType);
     }
 
-    @Override
-    public byte[] getMbusUserKey(String mbusDeviceIdentification) {
-        throw new NotImplementedException();
+    public void storeNewKeys(String deviceIdentification, Map<SecurityKeyType, byte[]> keysByType) {
+        this.validateKeys(keysByType);
+        TypedSecrets typedSecrets = new TypedSecrets();
+        List<TypedSecret> typedSecretList = typedSecrets.getTypedSecret();
+        for (Map.Entry<SecurityKeyType,byte[]> entry : keysByType.entrySet()) {
+            TypedSecret ts = new TypedSecret();
+            ts.setType(entry.getKey().toSecretType());
+            ts.setSecret(this.encryptSoapSecret(entry.getValue(),true));
+            typedSecretList.add(ts);
+        }
+        StoreSecretsRequest request = this.createStoreSecretsRequest(deviceIdentification, typedSecrets);
+        this.secretManagementClient.storeSecretsRequest(request);
     }
 
-    @Override
-    public byte[] getDlmsPassword(String deviceIdentification) {
-        throw new NotImplementedException();
+    private void validateKeys(Map<SecurityKeyType, byte[]> keysByType) {
+        long nrNulls = keysByType.values().stream().filter(Objects::isNull).count();
+        if (nrNulls > 0) {
+            throw new IllegalArgumentException(
+                    String.format("Provided %s keys, %s of which were NULL", keysByType.size(), nrNulls));
+        }
     }
 
-    @Override
-    public DlmsDevice storeNewKey(DlmsDevice device, byte[] encryptedKey, SecurityKeyType keyType) {
-        throw new NotImplementedException();
+    private StoreSecretsRequest createStoreSecretsRequest(String deviceIdentification, TypedSecrets typedSecrets) {
+        StoreSecretsRequest request = new StoreSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setTypedSecrets(typedSecrets);
+        return request;
     }
 
-    @Override
-    public DlmsDevice validateNewKey(DlmsDevice device, SecurityKeyType keyType) throws ProtocolAdapterException {
-        throw new NotImplementedException();
+    /**
+     * Updates the state of a new key from 'new' to 'active'
+     * <p>
+     * This method should be called to activate a new key stored with
+     * {@link #storeNewKeys(String, Map)} after it has
+     * been confirmed to be set on the device.
+     *
+     * @param deviceIdentification
+     *         DLMS device id
+     * @param keyType
+     *         type of key
+     *
+     * @throws ProtocolAdapterException
```

Comment:
Looking at the comment by Jenkins below, the `ProtocolAdapterException` is not actually thrown here (anymore).
Leaving the throws clause in the javadoc and declaration appears to be only misleading.